Source/WebCore/style/RuleSet.cpp

```diff
@@ -460,5 +458,25 @@ void RuleSet::shrinkToFit()
     m_resolverMutatingRulesInLayers.shrinkToFit();
 }
 
+Vector<RuleSet::ScopeQuery> RuleSet::scopeQueriesFor(const RuleData& ruleData) const
+{
+    if (m_scopeIdentifierForRulePosition.size() <= ruleData.position())
+        return { };
+
+    Vector<ScopeQuery> queries;
+
+    auto identifier = m_scopeIdentifierForRulePosition[ruleData.position()];
+    while (identifier) {
+        auto& query = m_scope[identifier - 1];
+        auto& scopeStart = query.scopeRule->scopeStart();
+        auto& scopeEnd = query.scopeRule->scopeEnd();
+        queries.append({ scopeStart, scopeEnd });
+        identifier = query.parent;
+    };
+
+    return queries;
```

Comment:
It is very inefficient to make copies of CSSSelectorLists in the middle of style resolution. I think this could just return Vector<const StyleRuleScope*>.

Reply:
`Vector<Ref<const StyleRuleScope>>` ?

Reply:
probably better!